src/yaca_conv.rs

```diff
@@ -0,0 +1,296 @@
+
+use crate::*;
+
+
+const BASE_ERROR_YACA:              c_int = -0x01E30000;
+const YACA_ERROR_NONE:              c_int = 0;
+const YACA_ERROR_INVALID_PARAMETER: c_int = -22;
+const YACA_ERROR_OUT_OF_MEMORY:     c_int = -12;
+const YACA_ERROR_INTERNAL:          c_int = BASE_ERROR_YACA | 0x01;
+const YACA_ERROR_DATA_MISMATCH:     c_int = BASE_ERROR_YACA | 0x02;
+const YACA_ERROR_INVALID_PASSWORD:  c_int = BASE_ERROR_YACA | 0x03;
+
+pub(crate) fn res_c_to_rs(r: c_int) -> crate::Result<()>
+{
+    match r {
+        YACA_ERROR_NONE => Result::Ok(()),
+        YACA_ERROR_INVALID_PARAMETER => Result::Err(Error::InvalidParameter),
+        YACA_ERROR_OUT_OF_MEMORY => Result::Err(Error::OutOfMemory),
+        YACA_ERROR_INTERNAL => Result::Err(Error::Internal),
+        YACA_ERROR_DATA_MISMATCH => Result::Err(Error::DataMismatch),
+        YACA_ERROR_INVALID_PASSWORD => Result::Err(Error::InvalidPassword),
+        x => Result::Err(Error::Unknown(x)),
+    }
+}
+
+pub(crate) fn res_c_to_rs_bool(r: c_int) -> crate::Result<bool>
+{
+    match r {
+        YACA_ERROR_NONE => Result::Ok(true),
+        YACA_ERROR_INVALID_PARAMETER => Result::Err(Error::InvalidParameter),
+        YACA_ERROR_OUT_OF_MEMORY => Result::Err(Error::OutOfMemory),
+        YACA_ERROR_INTERNAL => Result::Err(Error::Internal),
+        YACA_ERROR_DATA_MISMATCH => Result::Ok(false),
+        YACA_ERROR_INVALID_PASSWORD => Result::Err(Error::InvalidPassword),
+        x => Result::Err(Error::Unknown(x)),
+    }
+}
+
+const YACA_KEY_TYPE_SYMMETRIC:  c_int = 0;
+const YACA_KEY_TYPE_DES:        c_int = 1;
+const YACA_KEY_TYPE_IV:         c_int = 2;
+const YACA_KEY_TYPE_RSA_PUB:    c_int = 3;
+const YACA_KEY_TYPE_RSA_PRIV:   c_int = 4;
+const YACA_KEY_TYPE_DSA_PUB:    c_int = 5;
+const YACA_KEY_TYPE_DSA_PRIV:   c_int = 6;
+const YACA_KEY_TYPE_DH_PUB:     c_int = 7;
+const YACA_KEY_TYPE_DH_PRIV:    c_int = 8;
+const YACA_KEY_TYPE_EC_PUB:     c_int = 9;
+const YACA_KEY_TYPE_EC_PRIV:    c_int = 10;
+const YACA_KEY_TYPE_DSA_PARAMS: c_int = 11;
+const YACA_KEY_TYPE_DH_PARAMS:  c_int = 12;
+const YACA_KEY_TYPE_EC_PARAMS:  c_int = 13;
+
+
+pub(crate) fn key_type_rs_to_c(kt: &KeyType) -> c_int
+{
+    match kt {
+        KeyType::Symmetric => YACA_KEY_TYPE_SYMMETRIC,
+        KeyType::Des => YACA_KEY_TYPE_DES,
+        KeyType::Iv => YACA_KEY_TYPE_IV,
+        KeyType::Rsa(KeySubType::Public) => YACA_KEY_TYPE_RSA_PUB,
+        KeyType::Rsa(KeySubType::Private) => YACA_KEY_TYPE_RSA_PRIV,
+        KeyType::Dsa(KeySubType::Public) => YACA_KEY_TYPE_DSA_PUB,
+        KeyType::Dsa(KeySubType::Private) => YACA_KEY_TYPE_DSA_PRIV,
+        KeyType::Dh(KeySubType::Public) => YACA_KEY_TYPE_DH_PUB,
+        KeyType::Dh(KeySubType::Private) => YACA_KEY_TYPE_DH_PRIV,
+        KeyType::Ec(KeySubType::Public) => YACA_KEY_TYPE_EC_PUB,
+        KeyType::Ec(KeySubType::Private) => YACA_KEY_TYPE_EC_PRIV,
+        KeyType::Rsa(KeySubType::Params) => panic!("No KeySubType::Params for RSA"),
```

Comment:
How? Create a separate KeySubType for RSA only?

Reply:
Why not. Any way obviating the runtime check would work.

Reply:
Not sure I like it. Any other solution you had in mind?

Reply:
No preference here, any solution making ```panic!``` go works for me.

Reply:
And return EINVAL in such case? Ok.

Reply:
Why the preference to turn a compile time error into a runtime one?

Reply:
Because I don't like having 2 similar structs doing basically the same thing. Also I re-export KeySubType::* in yaca.
Please elabore fully what names would you assign in such case (having yaca::Private, yaca::Public, etc to make it look nice).

Reply:
They would be doing different things, one being effectively a superset of the other. If you don't like this, I don't feel there's any way I can convince you.

Let me just say why I like it:
* I dislike runtime errors that could be statically prevented.
* Longer names (like, say, dropping wildcard exports or introducing variants like ```RsaPublic``` or sth) don't bother me much. Frankly, I could live without variant wildcard exports altogether, even at the cost of noise, as long as I get a type hierarchy that accurately models the domain.

Reply:
> If you don't like this, I don't feel there's any way I can convince you.

I did not say there is no way to convince me. I asked you to propose the naming scheme that could convince me.

Reply:
How about losing ```KeySubType``` altogether and flattening prv/pub/par out into ```KeyType``` instead?

Reply:
The way it was originally... I kinda like that subtype thing. This is how I imagine it should be in Rust given it has such functionality for enums.

Reply:
Yeah I know but it seems to me like a siren's call. It's ostensibly beautiful but, in the end, the user has to name a concrete key type anyway. The difference between ```Rsa(Private)``` and ```RsaPrivate``` is superficial to these eyes.

Reply:
Aren't those a very tired, old eyes? ;-)

Reply:
Older than yours, true :) No worries though, I even took off my glasses for this one ;p

Reply:
Done